### PR TITLE
Allow collections of references to be stored in schemas

### PR DESCRIPTION
**Pull request by shans**

### CODE REVIEW

File: runtime/test/manifest-test.js [hunk 1]
```diff
@@ -1316,11 +1320,63 @@ resource SomeName
...
+        create as handle
+        P
+          bar = handle
+    `);
+
+    let [recipe] = manifest.recipes;
+    assert(recipe.normalize());
+    assert(recipe.isResolved());
+    const schema = recipe.particles[0].connections.bar.type.entitySchema;
+    const innerSchema = schema.fields.foo.schema.schema.model.entitySchema;
```
Comment (piotrswigon) [inline runtime/test/manifest-test.js]: Are you happy with schema.schema? It's not clear to me what they refer to exactly, but I'm leaving this to you.

File: runtime/test/reference-test.js [hunk 2]
```diff
@@ -358,4 +357,163 @@ describe('references', function() {
...
+    assert.equal(values[1].rawData.value, 'what another result!');
+  });
+
+  it('can construct collections of references in schemas', async () => {
+    let loader = new StubLoader({
+      manifest: `
+        schema Result
+          Text value
+        
+        particle ExtractReference in 'extractReference.js'
```
Comment (piotrswigon) [inline runtime/test/reference-test.js]: ExtractReference is a C&P from the above test, maybe update the name to something like "CreateReference" ?

File: runtime/test/schema-tests.js [hunk 3]
```diff
@@ -211,13 +211,28 @@ describe('schema', function() {
...
-    assert.doesNotThrow(() => {new References({one: new Reference({id: 'test', storageKey: 'test'}, Type.newEntity(ReferencedOneSchema), null), two: null}); });
-    assert.throws(() => {new References({one: null, two: new Reference({id: 'test', storageKey: 'test'}, Type.newEntity(ReferencedOneSchema), null)}); }, TypeError,
+    assert.doesNotThrow(() => {new References({one: new Reference({id: 'test', storageKey: 'test'}, Type.newReference(Type.newEntity(ReferencedOneSchema)), null), two: null}); });
+    assert.throws(() => {new References({one: null, two: new Reference({id: 'test', storageKey: 'test'}, Type.newReference(Type.newEntity(ReferencedOneSchema)), null)}); }, TypeError,
                   `Cannot set reference two with value '[object Object]' of mismatched type`);
     assert.throws(() => {new References({one: 42, two: null}); }, TypeError,
                   `Cannot set reference one with non-reference '42'`);
   });
 
+  it('collection types', async function() {
```
Comment (piotrswigon) [inline runtime/test/schema-tests.js]: describe the test better? I believe the test name should be a sentence e.g. it('can do some magical thing', ...)

maybe it('enforces the type of the collections of references' ?

**Comment by piotrswigon** [comment]
BTW awesome test coverage, thanks for doing the work!

File: runtime/ts/schema.ts [hunk 4]
```diff
@@ -300,25 +324,56 @@ export class Schema {
...
+            throw new TypeError(`Cannot set reference ${name} with non-reference '${value}'`);
+          }
+        } else if (type.kind === 'schema-collection' && value) {
+          // WTF?! value instanceof Set is returning false sometimes here because the Set in
+          // this environment (a native code constructor) isn't equal to the Set that the value
+          // has been constructed with (another native code constructor)...
+          if (value.constructor.name === 'Set') {
+            return value;
+          } else if (value.length && value instanceof Object) {
+            console.log('~~~~~', value.constructor, Set, value instanceof Set);
```
[the remark this answers is not in the document]
Comment (shans) [reply runtime/ts/schema.ts]: Done

File: runtime/test/manifest-test.js (hunk 1 above)
Comment (shans) [reply runtime/test/manifest-test.js]: Not that happy but changing it is a big job.. I'll put it in the list of technical debt.

File: runtime/test/reference-test.js (hunk 2 above)
Comment (shans) [reply runtime/test/reference-test.js]: Done.

File: runtime/test/schema-tests.js (hunk 3 above)
Comment (shans) [reply runtime/test/schema-tests.js]: Done.

**Pull request merged**